### PR TITLE
feat(headless): Refactor the Headless Notify Trigger controller to make it compatible with the insight use case

### PR DESCRIPTION
**Pull request by SimonMilord**
[SFINT-5468](https://coveord.atlassian.net/browse/SFINT-5468)

### IN THIS PR:

We want to refactor the Headless Notify trigger controller to make it compatible with the insight use case. We therefore will have the Core Triggers controller and have 2 controllers (1 for insight usecase and 1 for search usecase)

We are doing the following steps:

1. Created an insight-trigger-analytics-actions.ts file and added the `logNotifyTrigger` action in the file.
2. Added insight-trigger-analytics-actions.test.ts file to test the `logNotifyTrigger` action described above.
3. Modified the Core controller + tests
4. Added a insightTriggerController + tests
5. Modified the search trigger controller + tests
6. updated exports files in index

### DEMO IN HIP:

https://github.com/user-attachments/assets/25d25833-7083-45e4-ab2c-98d7638aeba5

### TESTS:

HEADLESS-CORE-NOTIFY-TRIGGER:
image

HEADLESS-NOTIFY-TRIGGER:
image

HEADLESS-INSIGHT-NOTIFY-TRIGGER:
image

INSIGHT-TRIGGER-ANALYTICS-ACTIONS:
image

[SFINT-5468]: https://coveord.atlassian.net/browse/SFINT-5468?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/headless/src/controllers/insight/triggers/headless-insight-notify-trigger.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export type {NotifyTrigger, NotifyTriggerState};
+
+/**
+ * Creates an insight `NotifyTrigger` controller instance.
+ *
+ * @param engine - The insight engine.
+ * @returns A `NotifyTrigger` controller instance.
+ * */
+export function buildNotifyTrigger(engine: InsightEngine): NotifyTrigger {
```
Comment (fbeaudoincoveo) [inline packages/headless/src/controllers/insight/triggers/headless-insight-notify-trigger.ts:17]: From what I can tell, the only thing that differs between the insight notify trigger controller and the search notify trigger controller is that they dispatch different analytics actions.

Instead of overriding the `subscribe` method, I think you could pass the analytics action creator as a prop to the core controller builder, and then just return the controller, i.e.,

```typescript
// headless-core-notify-trigger.ts

// ...
export interface NotifyTriggerProps {
  options: NotifyTriggerOptions
}

export interface NotifyTriggerOptions {
  logNotifyTriggerActionCreator: () => Action;
}

export function buildCoreNotifyTrigger(engine: CoreEngine, props: NotifyTriggerProps): NotifyTrigger {
  const {logNotifyTrigger} = props.options.logNotifyTriggerActionCreator;
  // ...
}
```

```typescript
// headless-insight-notify-trigger.ts

//...
import {logNotifyTrigger} from '../../../features/triggers/insight-trigger-analytics-actions';
// ...

export function buildNotifyTrigger(engine: InsightEngine): NotifyTrigger {
  return buildCoreNotifyTrigger(engine, {
    options: {
      logNotifyTriggerActionCreator: logNotifyTrigger,
    },
  });
}
```

```typescript
// headless-notify-trigger.ts

//...
import {logNotifyTrigger} from '../../features/triggers/trigger-analytics-actions';
// ...

export function buildNotifyTrigger(engine: SearchEngine): NotifyTrigger {
  return buildCoreNotifyTrigger(engine, {
    options: {
      logNotifyTriggerActionCreator: logNotifyTrigger,
    },
  });
}
```

File: packages/headless/src/features/triggers/insight-trigger-analytics-actions.test.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  firstSentences: 'first-sentences',
+  flags: 'flags',
+  rankingModifier: 'example rankingModifier',
+  raw: buildMockRaw({
+    urihash: 'example documentUriHash',
+    source: 'example sourceName',
+    collection: 'example collectionName',
+    permanentid: examplePermanentId,
+  }),
+};
```
Comment (mmitiche) [inline packages/headless/src/features/triggers/insight-trigger-analytics-actions.test.ts:46]: This seems unrelated to the notify triggers, why do we need it?

File: packages/headless/src/features/triggers/insight-trigger-analytics-actions.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import {SearchPageEvents} from '../analytics/search-action-cause';
+
+export const logNotifyTrigger = (): InsightAction =>
+  makeInsightAnalyticsActionFactory(SearchPageEvents.triggerQuery)(
+    'analytics/trigger/notify',
+    (client, state) => {
+      if (!state.triggers?.notifications.length) {
+        return null;
+      }
+      return client.logTriggerNotify({
```
Comment (mmitiche) [inline packages/headless/src/features/triggers/insight-trigger-analytics-actions.ts:14]: looking at the way you defined this new action in coveo.analytics.js, I see that this action also accepts case metadata, but you are not using them here

File: packages/headless/src/features/triggers/insight-trigger-analytics-actions.test.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          },
+          caseId: exampleCaseId,
+          caseNumber: exampleCaseNumber,
+        },
+        triggers: {
+          ...getTriggerInitialState(),
+          notifications: expectedNotifications,
+        },
+        search: buildMockSearchState({
+          results: [exampleResult],
```
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/headless/src/features/triggers/insight-trigger-analytics-actions.test.ts:80]: good point, I guess for the purpose of this test it is not needed, will remove.

File: packages/headless/src/features/triggers/insight-trigger-analytics-actions.ts (hunk 3 above)
Comment (SimonMilord) [reply packages/headless/src/features/triggers/insight-trigger-analytics-actions.ts:14]: right good point!

File: packages/headless/src/controllers/insight/triggers/headless-insight-notify-trigger.ts (hunk 1 above)
Comment (SimonMilord) [reply packages/headless/src/controllers/insight/triggers/headless-insight-notify-trigger.ts:17]: yeah good idea!

File: packages/headless/src/controllers/core/triggers/headless-core-notify-trigger.test.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
+import {logNotifyTrigger} from '../../../features/triggers/trigger-analytics-actions';
+import {triggerReducer as triggers} from '../../../features/triggers/triggers-slice';
+import {
+  buildMockSearchEngine,
+  MockedSearchEngine,
+} from '../../../test/mock-engine-v2';
+import {createMockState} from '../../../test/mock-state';
+import {NotifyTrigger} from '../../core/triggers/headless-core-notify-trigger';
+import {buildCoreNotifyTrigger} from './headless-core-notify-trigger';
```
Comment (fbeaudoincoveo) [inline packages/headless/src/controllers/core/triggers/headless-core-notify-trigger.test.ts:9]: I believe you'll have to add `.js` extensions to your relative imports, otherwise the build will fail

(same comment everywhere it applies)

**Review by mmitiche — COMMENTED** [review]
changes look good ✅

Please fix the build so we can make sure the unit tests are passing and we can approve and merge.

To fix the build you need to add the `.js` to the path of your imports

File: packages/headless/src/controllers/core/triggers/headless-core-notify-trigger.test.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  function setEngineTriggersState(notifications: string[]) {
+    engine.state.triggers!.notifications = notifications;
+  }
+
+  function registeredListeners() {
+    return (engine.subscribe as jest.Mock).mock.calls.map((args) => args[0]);
+  }
+
+  beforeEach(() => {
+    jest.resetAllMocks();
```
Comment (fbeaudoincoveo) [inline packages/headless/src/controllers/core/triggers/headless-core-notify-trigger.test.ts:34]: We're now using vitest instead of jest

The APIs are very similar, so the migration is straightforward (e.g., here I think you'd just have to do `vi.resetAllMocks();`)

**Review by mmitiche — APPROVED** [review]
LGTM!
